src/test/java/com/aws/greengrass/clientdevices/auth/infra/NetworkStateTest.java

```diff
@@ -0,0 +1,103 @@
+
+import com.aws.greengrass.mqttclient.CallbackEventManager;
+import com.aws.greengrass.mqttclient.MqttClient;
+import com.aws.greengrass.testcommons.testutilities.GGExtension;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import software.amazon.awssdk.crt.mqtt.MqttClientConnectionEvents;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.not;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.Mockito.verify;
+
+
+@ExtendWith({MockitoExtension.class, GGExtension.class})
+public class NetworkStateTest {
+    @Mock
+    MqttClient mqttClient;
+    @Captor
+    ArgumentCaptor<CallbackEventManager.OnConnectCallback> onConnectCaptor;
+    @Captor
+    ArgumentCaptor<MqttClientConnectionEvents> connectionEventsArgumentCaptor;
+    private ForkJoinPool fjp = new ForkJoinPool();
+    private NetworkState networkState;
+
+    @BeforeEach
+    void beforeEach() {
+        fjp = new ForkJoinPool();
+        networkState = new NetworkState(mqttClient, fjp);
+        verify(mqttClient).addToCallbackEvents(onConnectCaptor.capture(), connectionEventsArgumentCaptor.capture());
+    }
+
+    @AfterEach
+    void afterEach() {
+        fjp.shutdown();
+    }
+
+    void assertNetworkDomainEventOnNetworkEvent(Runnable networkEvent, NetworkState.ConnectionState expectedState)
+            throws InterruptedException {
+        CountDownLatch cdl = new CountDownLatch(1);
+        AtomicLong callbackThreadId = new AtomicLong();
+        AtomicReference<NetworkState.ConnectionState> connState = new AtomicReference<>();
+        long curThreadId = Thread.currentThread().getId();
+
+        Consumer<NetworkState.ConnectionState> networkStateConsumer = (state) -> {
+            connState.set(state);
+            callbackThreadId.set(Thread.currentThread().getId());
+            cdl.countDown();
+        };
+        networkState.registerHandler(networkStateConsumer);
+
+        networkEvent.run();
+
+        // Assert that NETWORK_UP event was emitted, and that it was emitted in a separate thread.
+        // We're checking this because network change events are triggered in a CRT thread, which
+        // we should not ever block
+        assertTrue(cdl.await(1, TimeUnit.SECONDS));
+        assertThat(callbackThreadId.get(), is(not(curThreadId)));
+        assertThat(connState.get(), is(expectedState));
+    }
+
```

Comment:
There is also one test missing. It seems from the code that the intent is that multiple handlers get ran by the same thread rather than trigger each handler on a single thread. If that is a decision we are firm on the we should have a test with more than one handler and check they both got triggered by the same thread